stdlib/int16.gr

```diff
@@ -0,0 +1,485 @@
+/**
+ * Utilities for working with the Int16 type.
+ * @example include "int16"
+ *
+ * @since v0.6.0
+ */
+
+module Int16
+
+include "runtime/unsafe/wasmi32"
+include "runtime/exception"
+include "runtime/numbers"
+from Numbers use { coerceNumberToShortInt, coerceShortIntToNumber }
+include "runtime/unsafe/constants"
+from Constants use { _SMIN16_I32, _SMAX16_I32, _SMIN16_I64, _SMAX16_I64 }
+
+@unsafe
+let _DATA_OFFSET = 8n
+@unsafe
+let _TAG_BYTE = 0b10010n
+@unsafe
+let _REMOVE_EXTRA_MASK = 0x00ffffffn
+@unsafe
+let _DATA_MASK = 0xffff00n
+
+/**
+ * @section Conversions: Functions for converting between Numbers and the Int16 type.
+ */
+
+/**
+ * Converts a Number to an Int16.
+ *
+ * @param number: The value to convert
+ * @returns The Number represented as an Int16
+ *
+ * @since v0.6.0
+ */
+@unsafe
+provide let fromNumber = (number: Number) => {
+  let val = coerceNumberToShortInt(
+    number,
+    _SMIN16_I32,
+    _SMAX16_I32,
+    _SMIN16_I64,
+    _SMAX16_I64,
+    false
+  )
+  let tagged = WasmI32.or(WasmI32.shl(val, _DATA_OFFSET), _TAG_BYTE)
+  WasmI32.toGrain(tagged): Int16
+}
+
+/**
+ * Converts an Int16 to a Number.
+ *
+ * @param value: The value to convert
+ * @returns The Int16 represented as a Number
+ *
+ * @since v0.6.0
+ */
+@unsafe
+provide let toNumber = (value: Int16) => {
+  coerceShortIntToNumber(
+    WasmI32.shrU(WasmI32.fromGrain(value), _DATA_OFFSET),
+    false
+  )
+}
+
+/**
+ * Converts an Uint16 to a Int16.
+ *
+ * @param number: The value to convert
+ * @returns The Uint16 represented as a Int16
+ * 
+ * @since v0.6.0
+ */
+@unsafe
+provide let fromUint16 = (x: Uint16) => {
+  let x = WasmI32.fromGrain(x)
+  // Trick: convert from Uint16 tag 100010 to Int16 tag 10010
+  let result = WasmI32.xor(x, 0b110000n)
+  WasmI32.toGrain(result): Int16
+}
+
+/**
+ * @section Operations: Mathematical operations for Int16 values.
+ */
+
+// Note: most of these operations do not attempt to correct trailing bits after
+// an operation for the sake of efficiency
```

Comment:
I think the 1 extra mask instruction is worth it to avoid a good amount of headache in the future. It'll also simplify the implementations of things like `equal` which can just do a simple check that the values are exactly the same, stuff like `compare` which can focus on simple checks, and simplify upcasts. Really removes a lot of the gotchas when we want to implement new stuff in the future.

Reply:
Sounds good

Reply:
Should signed ints have all leading zeros or should negative signed ints have leading 1s? The latter would make comparisons just require a normal i32 signed comparison but would make other operations slightly more expensive as they would require sign-extensions rather than simple masks

Reply:
Yeah, I think they should have leading ones. Kind of annoying, but probably worth it.